src/Products/Viewer/Controllers/ViewerApiController.cs

```diff
@@ -383,196 +411,318 @@ public HttpResponseMessage PrintPdf(PostedDataEntity loadDocumentRequest)
-            LoadDocumentEntity loadDocumentEntity = new LoadDocumentEntity();
-            // check if documentGuid contains path or only file name
-            if (!Path.IsPathRooted(documentGuid))
+            string password = (string.IsNullOrEmpty(postedData.password)) ? null : postedData.password;
+
+            string pageFilePathFormat;
+            ViewerSettings settings = GetViewerSettings(documentGuid, out pageFilePathFormat);
+
+            if (loadAllPages)
             {
-                documentGuid = globalConfiguration.Viewer.GetFilesDirectory() + "/" + documentGuid;
+                GenerateViewerCache(documentGuid, password, pageFilePathFormat, settings);
             }
-            dynamic documentInfoContainer;
-            // get document info options
-            DocumentInfoOptions documentInfoOptions = new DocumentInfoOptions();
-            // set password for protected document                
-            documentInfoOptions.Password = password;
-            // get document info container              
-            if (Path.GetExtension(documentGuid) == ".pdf" && globalConfiguration.Viewer.GetPrintAllowed())
+
+            return GetLoadDocumentEntity(documentGuid, password, settings, loadAllPages);
+        }
+
+        private void GenerateViewerCache(string documentGuid, string password, string pageFilePathFormat, ViewerSettings settings, int pageNumber = -1)
+        {
+            using (GroupDocs.Viewer.Viewer viewer = new GroupDocs.Viewer.Viewer(documentGuid, GetLoadOptions(password), settings))
             {
-                documentInfoContainer = this.GetHandler().GetDocumentInfo(documentGuid, documentInfoOptions) as PdfDocumentInfoContainer;
-                loadDocumentEntity.SetPrintAllowed(documentInfoContainer.PrintingAllowed);
+                if (globalConfiguration.Viewer.GetIsHtmlMode())
+                {
+                    HtmlViewOptions htmlViewOptions = HtmlViewOptions.ForEmbeddedResources(pageFilePathFormat);                  
+                    SetWatermarkOptions(htmlViewOptions);
+
+                    if (pageNumber < 0)
+                    {
+                        viewer.View(htmlViewOptions);
+                    }
+                    else
+                    {
+                        viewer.View(htmlViewOptions, pageNumber);
+                    }
+                }
+                else
+                {
+                    PngViewOptions pngViewOptions = new PngViewOptions();
+                    SetWatermarkOptions(pngViewOptions);
+
+                    if (pageNumber < 0)
+                    {
+                        viewer.View(pngViewOptions);
+                    }
+                    else
+                    {
+                        viewer.View(pngViewOptions, pageNumber);
+                    }
+                }
             }
-            else
+        }
+
+        private ViewerSettings GetViewerSettings(string documentGuid, out string pageFilePathFormat)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Make 'GetViewerSettings' a static method.](https://app.codacy.com/gh/groupdocs-viewer/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5242614)